src/main/java/net/minestom/server/instance/block/predicate/DataComponentPredicate.java

```diff
@@ -0,0 +1,376 @@
+                "levels", DataComponentPredicates.INT_RANGE_CODEC.optional(), Enchantment::levels,
+                Enchantment::new
+        );
+
+        public boolean test(DataComponent.Holder holder) {
+            EnchantmentList holderEnchants = holder.get(DataComponents.ENCHANTMENTS);
+            if (holderEnchants == null) {
+                return false;
+            }
+            if (enchantments == null) {
+                // If `enchantments` is not specified, the predicate returns true when any enchantment matches the specified `level`
+                return holderEnchants.enchantments().entrySet().stream().anyMatch(entry -> levels == null || levels.inRange(entry.getValue()));
+            }
+            for (DynamicRegistry.Key<net.minestom.server.item.enchant.Enchantment> e : enchantments) {
+                if (holderEnchants.has(e) && (levels == null || levels.inRange(holderEnchants.level(e)))) {
+                    return true;
+                }
+            }
+
+            return false;
+        }
+    }
+
+    record StoredEnchantments(List<StoredEnchantment> children) implements DataComponentPredicate {
+        public static Codec<StoredEnchantments> CODEC = StoredEnchantment.CODEC.list().transform(StoredEnchantments::new, StoredEnchantments::children);
+
+        @Override
+        public boolean test(DataComponent.Holder holder) {
+            return children.stream().allMatch(enchantment -> enchantment.test(holder));
+        }
+    }
+
+    record StoredEnchantment(
+            @Nullable List<DynamicRegistry.Key<net.minestom.server.item.enchant.Enchantment>> enchantments,
+            Range.Int levels) {
+
+        public static Codec<StoredEnchantment> CODEC = StructCodec.struct(
+                "enchantments", net.minestom.server.item.enchant.Enchantment.CODEC.listOrSingle().optional(), StoredEnchantment::enchantments,
+                "levels", DataComponentPredicates.INT_RANGE_CODEC.optional(), StoredEnchantment::levels,
+                StoredEnchantment::new
+        );
+
+        public boolean test(DataComponent.Holder holder) {
+            EnchantmentList holderEnchants = holder.get(DataComponents.STORED_ENCHANTMENTS);
+            if (holderEnchants == null) {
+                return false;
+            }
+            if (enchantments == null) {
+                // If `enchantments` is not specified, the predicate returns true when any enchantment matches the specified `level`
+                return holderEnchants.enchantments().entrySet().stream().anyMatch(entry -> levels == null || levels.inRange(entry.getValue()));
+            }
+            for (DynamicRegistry.Key<net.minestom.server.item.enchant.Enchantment> e : enchantments) {
+                if (holderEnchants.has(e) && (levels == null || levels.inRange(holderEnchants.level(e)))) {
+                    return true;
+                }
+            }
+
+            return false;
+        }
+    }
```

Comment:
You should be able to deduplicate this by having `Enchantments`/`StoredEnchantments` pass in the relevant `EnchantmentList`. There don't need to be two implementations.